### ISSUE TITLE
P2P replication error (lack of Digest auth support in CBLChangeTracker)

### ISSUE THREAD

**Issue by joshblour** [couchbase/couchbase-lite-ios#777]
I'm not able to replicate from a CBLListener with authentication turned on. Works great with authentication turned off.

CBLListener code:
```` objc
self.listener = [[CBLListener alloc] initWithManager:[CBLManager sharedInstance] port:52021];
self.listener.requiresAuth = YES;
[self.listener setPasswords:@{[self authenticatorUsername] : [self authenticatorPassword]}];
[self.listener setBonjourName:[self myUUID] type: [self bonjourType]];
[self.listener start:&error]
````
Client code:
```objc
NSURLComponents *components = [[NSURLComponents alloc] init];
components.scheme = @"http";
components.host = service.hostName;
components.port = @(service.port);
components.path = [NSString stringWithFormat:@"/%@", [[self database] name]];
components.user = [self authenticatorUsername];
components.password = [self authenticatorPassword];
self.pullReplication = [[self database] createPullReplication:components.URL];
self.pullReplication.continuous = self.isContinuous;
self.pullReplication.customProperties = @{@"websocket": @NO};
self.pullReplication.filter = @"exceptUUIDAndMinUpdatedAt";
self.pullReplication.filterParams = @{@"exceptUUID": uuid, @"minUpdatedAt": [CBLJSON JSONObjectWithDate:minUpdatedAt]};
[self.pullReplication start];
```

Logs from the client here: https://gist.github.com/joshblour/59b76b2531dd0f2eb3b4
Logs from the CBLListener here: https://gist.github.com/joshblour/163a9d101f1f4b10dc56

The error seems to happen on the CBLListener side: 
````
CBLListener: Login attempted for user 'p2p_v1'
CBLListener: GET /proximator/_local/e95ce455ab12e1261e4f6b592ee4cc8f478c8eaa
CBLListener:     Response[GET /proximator/_local/e95ce455ab12e1261e4f6b592ee4cc8f478c8eaa] --> 404
````

**Comment by joshblour** [comment]
I'm able to replicate this in my fork of the Grocery Sync app: 
https://github.com/joshblour/Grocery-Sync-iOS/tree/password

**Comment by snej** [comment]
FYI, the 404 status when GETting a `_local/` document is not an error. Those documents are replication checkpoints, and the 404 just means the server/peer doesn't yet have a saved checkpoint for this replication.

**Comment by snej** [comment]
The error on the client side:
```
12:20:12.444‖ CBLSocketChangeTracker[0x7ffc0c856800 proximator]: Connection error #1, retrying in 2.0 sec: The operation couldn’t be completed. (NSURLErrorDomain error -1005.)
```
-1005 is kCFURLErrorNetworkConnectionLost, i.e. the other side closed the socket abruptly.

**Comment by snej** [comment]
The log you showed on the listener side indicates that HTTP auth is succeeding and the client is able to make requests. So it looks like the problem is with the changes feed (CBLSocketChangeTracker.)

**Comment by pasin** [comment]
I could reproduce the issue with the Grocery Sync app that @joshblour provided. Need to debug the issue.

**Comment by pasin** [comment]
Here are some updates:

It seems like CBLSocketChangeTracker cannot do Digest Authentication. The authenticator set to the CBLSocketChangeTracker is a Basic Authenticator so the Basic Authorization header was set to the http message, but somehow the response code returned was -1005 instead of 401 or 407.

We don't have a specific authenticator for Digest authentication. Based on CBL_Replicator and CBLRemoteRequest implementation, we rely on receiving a challenge (401) returned by the server to do digest authentication. I'm not familiar with CFNetwork, but I guessed we are trying to do similar things. I would need to learn more about CFNetwork and challenge to know more what is going on.

**Comment by snej** [comment]
Hm … two things:

1. In -[CBLSocketChangeTracker start], the authorizer shouldn't be adding its headers to the message if it's a basic authorizer — that will leak the password over the net (and also it seems to trigger the -1005 error from the listener on the other end.)
2. WebSocketHTTPLogic's call to CFHTTPMessageAddAuthentication should be passing NULL instead of kCFHTTPAuthenticationSchemeBasic; that will allow it to use digest auth when appropriate.

It's a bit trickier than that because if we don't do 1 then we have to pass the authorizer's credential to the Logic...

**Comment by pasin** [comment]
I have verified that the issue was fixed by ad8bea4357eca523f912c757bc85d942f475f073, 4e9b06dd590976d0d02481a6c06f1537e74b348d, 34fc4a78ddfafa18b0ecbff3b9a30f39d4c7f59f.
